From ff07a06026be0deddd4ad6a9f89c9b6f6d1175c3 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Wed, 21 Mar 2012 09:23:07 +0100
Subject: [PATCH 1/4] scsi: do not send MODE SENSE except to QEMU disks

RH-Author: Paolo Bonzini <pbonzini@redhat.com>
Message-id: <1332321790-12045-2-git-send-email-pbonzini@redhat.com>
Patchwork-id: 38767
O-Subject: [RHEL6.3 PATCH seabios 1/4] scsi: do not send MODE SENSE except to QEMU disks
Bugzilla: 801293
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>

This is the simplest way to avoid breaking boot on USB sticks that
stall when asked for the MODE SENSE page 4.  Some old sticks do
not support the MODE SENSE command at all and just return a
"medium may have changed" unit attention condition when SeaBIOS
sends it!

Reported-by: Dave Frodin <dave@camp.se-eng.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry-picked from commit cb721714570520c02ae48efc26d3c04b8548d973)
---
 src/blockcmd.c |   37 +++++++++++++++++++++++--------------
 1 files changed, 23 insertions(+), 14 deletions(-)

Signed-off-by: Michal Novotny <minovotn@redhat.com>
---
 src/blockcmd.c |   37 +++++++++++++++++++++++--------------
 1 files changed, 23 insertions(+), 14 deletions(-)

diff --git a/src/blockcmd.c b/src/blockcmd.c
index 9919acb..4f3131c 100644
--- a/src/blockcmd.c
+++ b/src/blockcmd.c
@@ -131,20 +131,29 @@ scsi_init_drive(struct drive_s *drive, const char *s, int *pdt, char **desc)
     dprintf(1, "%s blksize=%d sectors=%d\n"
             , s, drive->blksize, (unsigned)drive->sectors);
 
-    struct cdbres_mode_sense_geom geomdata;
-    ret = cdb_mode_sense_geom(&dop, &geomdata);
-    if (ret == 0) {
-        u32 cylinders;
-        cylinders = geomdata.cyl[0] << 16;
-        cylinders |= geomdata.cyl[1] << 8;
-        cylinders |= geomdata.cyl[2];
-        if (cylinders && geomdata.heads &&
-            drive->sectors <= 0xFFFFFFFFULL &&
-            ((u32)drive->sectors % (geomdata.heads * cylinders) == 0)) {
-            drive->pchs.cylinders = cylinders;
-            drive->pchs.heads = geomdata.heads;
-            drive->pchs.spt = (u32)drive->sectors
-     / (geomdata.heads * cylinders);
+    // We do not recover from USB stalls, so try to be safe and avoid
+    // sending the command if the (obsolete, but still provided by QEMU)
+    // fixed disk geometry page may not be supported.
+    //
+    // We could also send the command only to small disks (e.g. <504MiB)
+    // but some old USB keys only support a very small subset of SCSI which
+    // does not even include the MODE SENSE command!
+    //
+    if (! CONFIG_COREBOOT && memcmp(vendor, "QEMU    ", 8) == 0) {
+        struct cdbres_mode_sense_geom geomdata;
+        ret = cdb_mode_sense_geom(&dop, &geomdata);
+        if (ret == 0) {
+            u32 cylinders;
+            cylinders = geomdata.cyl[0] << 16;
+            cylinders |= geomdata.cyl[1] << 8;
+            cylinders |= geomdata.cyl[2];
+            if (cylinders && geomdata.heads &&
+                drive->sectors <= 0xFFFFFFFFULL &&
+                ((u32)drive->sectors % (geomdata.heads * cylinders) == 0)) {
+                drive->pchs.cylinders = cylinders;
+                drive->pchs.heads = geomdata.heads;
+                drive->pchs.spt = (u32)drive->sectors / (geomdata.heads * cylinders);
+            }
         }
     }
 
-- 
1.7.7.6

